IS0402Test.py

```diff
@@ -507,31 +412,595 @@ def test_20(self):
+        self.check_paged_response(test, self.do_paged_request(label = "bar", until = ts[12], limit = 3),
+                                  expected_ids = [ids[7], ids[8], ids[12]],
+                                  expected_since = ts[3], expected_until = ts[12], expected_limit = 3)
+
+        # Query 7: like Query 5, with paging.since specified, but still enough matches
+        #          filter         -, -, 2, 3, -, -, -, 7, 8, -, --, --, 12, 13, --, --, --, 17, 18, --
+        #          request                     (                           ]
+        #          response                    (       ^  ^              ^ ]
+
+        self.check_paged_response(test, self.do_paged_request(label = "bar", since = ts[4], until = ts[12], limit = 3),
+                                  expected_ids = [ids[7], ids[8], ids[12]],
+                                  expected_since = ts[4], expected_until = ts[12], expected_limit = 3)
+
+        # Query 8: like Query 5, with paging.since specified, and not enough matches
+        #          filter         -, -, 2, 3, -, -, -, 7, 8, -, --, --, 12, 13, --, --, --, 17, 18, --
+        #          request                                    (            ]
+        #          response                                   (          ^ ]
+
+        self.check_paged_response(test, self.do_paged_request(label = "bar", since = ts[9], until = ts[12], limit = 3),
+                                  expected_ids = [ids[12]],
+                                  expected_since = ts[9], expected_until = ts[12], expected_limit = 3)
+
+        # Query 9: like Query 5, but no matches
+        #          filter         -, -, 2, 3, -, -, -, 7, 8, -, --, --, 12, 13, --, --, --, 17, 18, --
+        #          request                                    (        ]
+        #          response                                   (        ]
+
+        self.check_paged_response(test, self.do_paged_request(label = "bar", since = ts[9], until = ts[11], limit = 3),
+                                  expected_ids = [],
+                                  expected_since = ts[9], expected_until = ts[11], expected_limit = 3)
+
+        return test.PASS()
+
+    def test_21_6(self):
+        """Query API implements pagination (bad requests)"""
+
+        test = Test("Query API implements pagination (bad requests)")
+        self.check_paged_trait(test)
+        description = "test_21_6"
+
+        before = self.is04_query_utils.get_TAI_time()
+        after = self.is04_query_utils.get_TAI_time(1)
+
+        # Specifying since after until is a bad request
+        valid, response, query_string = self.do_paged_request(since = after, until = before)
+
+        if not valid:
+            raise NMOSTestException(test.FAIL("Query API did not respond as expected, "
+                                              "for query: {}".format(query_string)))
+        elif response.status_code != 400:
```

Comment:
Does this one also need a 501 check?

Reply:
Thanks. Yes. I'll hack our Registry to return 501 so I'll pick this and any similar misses up.

Reply:
Resolved by 3837909c27935738d5afe3db8d9be1a73f485394